### PR TITLE
NEW: Add Feature Model Nuclei Instance Segmentation

### PR DESCRIPTION
**Pull request by vqdang**
- [x] Add HoVerNet Architecture and associated post-processing methods
- [x] Add Nucleus Instance Segmentor to specifically deal with nucleus instance segmentation on WSI levels
- [x] Add metric function to assess detection quality.
- - [x] F1 detection basing matching point sets.
- [x] Add function to overlay instance prediction onto canvas or image.
- [x] Add Tests. 

May need another PR to actually improve the doc and better explain the post-processing scheme.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -19,5 +19,7 @@ torchvision==0.10.1
 torch==1.9.1
 tqdm==4.60.0
 jupyterlab
+shapely
 requests
+pygeos
```
Comment (shaneahmed) [inline requirements.txt:24]: need to rearrange in alphabetical order.

File: requirements.win64.conda.yml [hunk 2]
```diff
@@ -29,6 +29,9 @@ dependencies:
   - tifffile
   - torchvision
   - tqdm
+  - jupyterlab
```
Comment (shaneahmed) [inline requirements.win64.conda.yml:32]: need to rearrange in alphabetical order.

File: tests/models/__init__.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# The Original Code is Copyright (C) 2021, TIALab, University of Warwick
```
Comment (shaneahmed) [inline tests/models/__init__.py:17]: TIA Centre

File: tiatoolbox/models/architecture/utils.py [hunk 4]
```diff
@@ -21,37 +21,90 @@
...
+from typing import Union
+
 import numpy as np
 import torch
 import torch.nn as nn
 
 
-def crop_op(x, cropping, data_format="NCHW"):
-    """Center crop image with substracted amount.
+def center_crop(
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:31]: use British or US English not a mix of both.
I think we are using British spellings so use centre then

File: tests/models/test_arch_utils.py [hunk 5]
```diff
@@ -24,7 +24,11 @@
 import pytest
 import torch
 
-from tiatoolbox.models.architecture.utils import UpSample2x, crop_op
+from tiatoolbox.models.architecture.utils import (
+    UpSample2x,
```
Comment (shaneahmed) [inline tests/models/test_arch_utils.py:28]: need to run isort I guess. Please use pre-commit hooks.

File: tests/models/test_nucleus_instance_segmentor.py [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+            [1, 1, 1, 1],
+        ]
+    )
+    boxes, flag = info[3]
+    assert np.sum(boxes - _boxes) == 0, "Wrong Cross Section Bounds"
+    assert np.sum(flag - _flag) == 0, "Fail Cross Section Flag"
+
+
+def test_crash_segmentor(remote_sample, tmp_path):
+    """Test crash."""
```
Comment (shaneahmed) [inline tests/models/test_nucleus_instance_segmentor.py:213]: A little more info in the docstring

File: tests/test_utils.py [hunk 7]
```diff
@@ -47,6 +47,14 @@ def test_imresize():
     )
     assert resized_img.shape == (1000, 500, 3)
 
+    # test for dtype conversion
```
Comment (shaneahmed) [inline tests/test_utils.py:50]: the assert should also check dtype then?

File: tiatoolbox/models/architecture/utils.py [hunk 8]
```diff
@@ -21,37 +21,90 @@
...
+    crop_l = crop_shape[1] // 2
+    crop_r = crop_shape[1] - crop_l
+    if data_format == "NCHW":
+        img = img[:, :, crop_t:-crop_b, crop_l:-crop_r]
+    else:
+        img = img[:, crop_t:-crop_b, crop_l:-crop_r, :]
+    return img
+
+
+def center_crop_to_shape(
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:62]: same as above, use British or US English.

File: tiatoolbox/utils/transforms.py [hunk 9]
```diff
@@ -111,6 +111,10 @@ def imresize(img, scale_factor=None, output_size=None, interpolation="optimise")
             interpolation = "cubic"
         else:
             interpolation = "area"
+
+    if np.float16 == img.dtype:
```
Comment (shaneahmed) [inline tiatoolbox/utils/transforms.py:115]: I would write it as `img.dtype==np.float16`

**Review by shaneahmed — CHANGES_REQUESTED** [review]
Thanks @vqdang 
Please address the changes and also add the new modules to documentation on readthedocs.

File: tiatoolbox/models/__init__.py [hunk 10]
```diff
@@ -20,3 +20,17 @@
 
 """Models package for the models implemented in tiatoolbox."""
 from tiatoolbox.models import abc, architecture, dataset
+from tiatoolbox.models.controller.nucleus_instance_segmentor import (
```
Comment (mostafajahanifar) [inline tiatoolbox/models/__init__.py:23]: shouldn't we go with `tiatoolbox.models.engin` instead of `controller` in this PR?

Comment (vqdang) [reply tiatoolbox/models/__init__.py:23]: That requires changing everything now too btw. Srijay is doing that atm for `develope`. Either we merge this one first and deal with the renaming later, or we stall till Srijay PR gets merged and then I deal with this.

@shaneahmed what do you think?

Comment (shaneahmed) [reply tiatoolbox/models/__init__.py:23]: Let's wait for @Srijay-lab PR. I think his PR should be quick to merge.

File: tiatoolbox/utils/visualization.py [hunk 11]
```diff
@@ -138,3 +161,45 @@ def overlay_patch_prediction(
     cbar.ax.tick_params(labelsize=12)
 
     return ax
+
+
+def overlay_instance_prediction(
```
Comment (mostafajahanifar) [inline tiatoolbox/utils/visualization.py:166]: As a minor suggestion, isn't it better to call this function: `overlay_prediction_contours` and the other one `overlay_prediction_masks` (instead of `overlay_patch_prediction`), or something general like these? because these two function can be used for other tasks as well. For example, we used `overlay_patch_prediction` function for depicting wsi semantic segmentation performance. On the other hand, `overlay_instance_prediction` can also be used to overlay of the contours of semantic segmentation results where `instance` term does not apply.

**Review by mostafajahanifar — APPROVED** [review]
I have made some minor suggestions which are mostly about documentations. Otherwise, you have done well @vqdang. Thank you very much.

File: tiatoolbox/models/architecture/hovernet.py [hunk 12]
```diff
@@ -0,0 +1,659 @@
...
+            pad_val_end = pad - pad_val_start
+            padding = (pad_val_start, pad_val_end, pad_val_start, pad_val_end)
+        x = F.pad(x, padding, "constant", 0)
+        return x
+
+
+class DenseBlock(nn.Module):
+    """Dense Block as defined in:
+
+    This will only performs `valid` convolution.
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/models/architecture/hovernet.py:79]: Not necessarily, resnet in torchvision doesnt have that but something very complicated like transformer will have.
Resnet
https://github.com/pytorch/vision/blob/b1fc290355176feedc8dde08edb049022d646cc0/torchvision/models/resnet.py#L164

Transformer
https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html

File: tiatoolbox/models/controller/nucleus_instance_segmentor.py [hunk 13]
```diff
@@ -0,0 +1,665 @@
...
+    IOSegmentorConfig,
+    SemanticSegmentor,
+    WSIStreamDataset,
+)
+from tiatoolbox.tools.patchextraction import PatchExtractor
+
+
+# till the day Python could natively pickle Object method/static method
+# the only way to passing function is top-level. Or we have to jump to
+# another third party solution
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/models/controller/nucleus_instance_segmentor.py:43]: Honestly I will just leave it there as a reminder untile someone resolve this.

File: tiatoolbox/utils/visualization.py (hunk 11 above)
Comment (vqdang) [reply tiatoolbox/utils/visualization.py:166]: Updated, you will have to change the notebook.

File: tiatoolbox/models/controller/nucleus_instance_segmentor.py [hunk 14]
```diff
@@ -0,0 +1,665 @@
...
+            inst_info["contour"] += tile_tl
+            inst_uuid = uuid.uuid4().hex
+            new_inst_dict[inst_uuid] = inst_info
+
+    return new_inst_dict, remove_insts_in_orig
+
+
+class NucleusInstanceSegmentor(SemanticSegmentor):
+    """Nucleus Instance Segmentor."""
+
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/models/controller/nucleus_instance_segmentor.py]: Doc added

File: tests/models/test_hovernet.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    model = HoVerNet(num_types=6, mode="fast")
+    fetch_pretrained_weights("hovernet_fast-pannuke", f"{tmp_path}/weigths.pth")
+    pretrained = torch.load(f"{tmp_path}/weigths.pth")
+    model.load_state_dict(pretrained)
+    output = model.infer_batch(model, batch, on_gpu=False)
+    output = [v[0] for v in output]
+    output = model.postproc(output)
+    assert len(output[1]) > 0, "Must have some nuclei."
+
+    # * test original mode
```
Comment (simongraham) [inline tests/models/test_hovernet.py:58]: Can you state what original mode is

File: tiatoolbox/utils/metrics.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
+# ***** BEGIN GPL LICENSE BLOCK *****
+#
```
Comment (simongraham) [inline tiatoolbox/utils/metrics.py:2]: Do we instead want to have tiatoolbox.metrics ? I am okay for this PR but we may want to consider a refactor. How do they do it in SKlearn ?

File: tiatoolbox/utils/visualization.py [hunk 17]
```diff
@@ -138,3 +161,45 @@ def overlay_patch_prediction(
     cbar.ax.tick_params(labelsize=12)
 
     return ax
+
+
+def overlay_prediction_contours(
```
Comment (simongraham) [inline tiatoolbox/utils/visualization.py:166]: I actually don’t like the random colours. I think if no type is supplied then we should default to yellow overlay

File: tiatoolbox/utils/visualization.py [hunk 18]
```diff
@@ -138,3 +161,45 @@ def overlay_patch_prediction(
...
+        canvas (ndarray): Image to draw predictions on.
+        inst_dict (dict): Dictionary of instances. It is expected to be
+            in the following format:
+            {instance_id: {type: int, contour: List[List[int]], centroid:List[float]}.
+        draw_dot: To draw a dot for each centroid or not.
+        type_colour: A dict of {type_id : (type_name, colour)},
+            `type_id` is from 0-N and `colour` is a tuple of (R, G, B).
+        line_thickness: line thickness of contours.
+
+    Return:
```
Comment (simongraham) [inline tiatoolbox/utils/visualization.py:181]: Returns (plural)

File: tiatoolbox/models/engine/nucleus_instance_segmentor.py [hunk 19]
```diff
@@ -0,0 +1,704 @@
...
+
+from tiatoolbox.models.engine.semantic_segmentor import (
+    IOSegmentorConfig,
+    SemanticSegmentor,
+    WSIStreamDataset,
+)
+from tiatoolbox.tools.patchextraction import PatchExtractor
+
+
+# till the day Python could natively pickle Object method/static method
```
Comment (simongraham) [inline tiatoolbox/models/engine/nucleus_instance_segmentor.py:41]: Can we make this comment shorter - it’s a bit long winded

File: tiatoolbox/models/engine/nucleus_instance_segmentor.py [hunk 20]
```diff
@@ -0,0 +1,704 @@
...
+            The expected format is {instance_id: {type: int,
+            contour: List[List[int]], centroid:List[float], box:List[int]}.
+        remove_insts_in_orig (list): List of instance id within `ref_inst_dict`
+            to be removed to prevent overlapping predictions. These instances
+            are those get cutoff at the boundary due to the tiling process.
+
+    """
+    locations, predictions = list(zip(*tile_output))
+
+    # convert from WSI space to Tile space
```
Comment (simongraham) [inline tiatoolbox/models/engine/nucleus_instance_segmentor.py:103]: Tile should be all lowercase

File: tiatoolbox/models/engine/nucleus_instance_segmentor.py [hunk 21]
```diff
@@ -0,0 +1,704 @@
...
+        head_raw = merge_predictions(
+            head_tile_shape[::-1],
+            head_predictions,
+            head_locations,
+            free_prediction=True,
+        )
+        head_raws.append(head_raw)
+    _, inst_dict = postproc(head_raws)
+
+    # should be extremely rare
```
Comment (simongraham) [inline tiatoolbox/models/engine/nucleus_instance_segmentor.py:133]: Explain what this means for clarity

File: tiatoolbox/models/engine/nucleus_instance_segmentor.py [hunk 22]
```diff
@@ -0,0 +1,704 @@
...
+            inst_info["centroid"] += tile_tl
+            inst_info["contour"] += tile_tl
+            inst_uuid = uuid.uuid4().hex
+            new_inst_dict[inst_uuid] = inst_info
+
+    return new_inst_dict, remove_insts_in_orig
+
+
+class NucleusInstanceSegmentor(SemanticSegmentor):
+    """An engine specifically designed to handle WSI inference.
```
Comment (simongraham) [inline tiatoolbox/models/engine/nucleus_instance_segmentor.py:258]: What about tiles ? We need to be able to apply functionality for that too as it will be very commonly required

File: tiatoolbox/models/engine/nucleus_instance_segmentor.py [hunk 23]
```diff
@@ -0,0 +1,704 @@
...
+
+        We modify the shared space so that we can update worker info without
+        needing to re-create the worker. There should be no race-condition
+        because only by looping `self._loader` in main thread will trigger querying
+        new data from each worker, and this portion should still be in sequential
+        execution order in the main thread.
+
+        Args:
+            wsi_idx (int): The index of the WSI to be processed. This is used to
+                to retrieve the file path.
```
Comment (simongraham) [inline tiatoolbox/models/engine/nucleus_instance_segmentor.py:509]: Repeat of word ‘to’

**Review by simongraham — CHANGES_REQUESTED** [review]
Nice work. Overall it is looking good - mainly comments to do with docs. Also, please clarify how easy it is to function tiles in addition to wsi ? Ideally we need to be able to allow this with the addition of an argument like we do in patch predictor. Getting the tile level results will be a requirement. If it is incorporated then it needs to be made clear as it seems like it is all geared towards wsi at the moment.

File: tiatoolbox/models/architecture/hovernet.py [hunk 24]
```diff
@@ -0,0 +1,677 @@
...
+          By default, no typing (`num_types=None`) is used.
+        mode (str): To use architecture defined in as in original paper
+          (`original`) or the one used in Pannuke (`fast`).
+
+    Reference:
+        Graham, Simon, et al. "Hover-net: Simultaneous segmentation
+        and classification of nuclei in multi-tissue histology images."
+        Medical Image Analysis 58 (2019): 101563.
+
+        Jevgenij, Gamper, et al. "PanNuke Dataset Extension, Insights
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/models/architecture/hovernet.py:296]: We also use `fast` mode here which was first used by Pannuke so we think we need to reference it.

File: tiatoolbox/models/architecture/hovernet.py [hunk 25]
```diff
@@ -0,0 +1,677 @@
...
+
+        Args:
+            imgs (torch.Tensor): Input images, the tensor is in the shape of NCHW.
+
+        Returns:
+            output (dict): A dictionary containing the inference output.
+                The expected format os {decoder_name: prediction}.
+
+        """
+        imgs = imgs / 255.0  # to 0-1 range to match XY
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/models/architecture/hovernet.py:417]: I'm happy to let it there, we assumed it fixed within HoVerNet like this anyways.

File: tiatoolbox/utils/metrics.py (hunk 16 above)
Comment (vqdang) [reply tiatoolbox/utils/metrics.py:2]: Kept under a folder but being exposed under `sklearn.metrics` namespace. I think we should just leave them here for now.

https://github.com/scikit-learn/scikit-learn/tree/0d378913be6d7e485b792ea36e9268be31ed52d0/sklearn/metrics

File: tiatoolbox/models/engine/nucleus_instance_segmentor.py (hunk 19 above)
Comment (vqdang) [reply tiatoolbox/models/engine/nucleus_instance_segmentor.py:41]: Not possible for me, you can paraphrase the following if you want. I think its the most compact I can write
```
# Python has yet being able to natively pickle Object method/static method.
# Only top-level function is passable to multi-processing as caller.
# May need 3rd party libraries to use method/static method otherwise.
```

File: tiatoolbox/models/architecture/hovernet.py [hunk 26]
```diff
@@ -0,0 +1,677 @@
...
+                        bias=False,
+                    ),
+                ),
+            ]
+            # * has bna to conclude each previous block so
+            # * must not put preact for the first unit of this block
+            unit_layer = unit_layer if idx != 0 else unit_layer[2:]
+            self.units.append(nn.Sequential(OrderedDict(unit_layer)))
+            unit_in_ch = unit_chs[-1]
+
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/models/architecture/hovernet.py:251]: Not even sure what you want to add here as It looks pretty straightforward to me

File: tiatoolbox/models/engine/nucleus_instance_segmentor.py (hunk 19 above)
Comment (simongraham) [reply tiatoolbox/models/engine/nucleus_instance_segmentor.py]: # Python is yet to be able to natively pickle Object method/static method.
# Only top-level function is passable to multi-processing as caller.
# May need 3rd party libraries to use method/static method otherwise.

File: tiatoolbox/models/architecture/hovernet.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (simongraham) [reply tiatoolbox/models/architecture/hovernet.py:251]: Let's keep it like it is for now and we can add some extra comments in models in another PR

**Review by shaneahmed — APPROVED** [review]

**Review by simongraham — APPROVED** [review]
